### PR TITLE
Perf optimization for conv and gemm kernels.

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37829 Some TODO fixes.
* **#37626 Perf optimization for conv and gemm kernels.**
* #37624 Added per channel separate test cases for fc and deconv tests.
* #37623 Changes to enable per channel support on dynamic linear.
* #37622 Enabled per channel quantized static linear/conv
* #37621 Added per channel kernels for depthwise conv.
* #37620 Changes to enable per channel requant.
* #37619 Enable per channel zero point.
* #37618 Interface changes to enable per channel quant.

Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Benchmarking was done via q8gemm-bench and convolution-bench on pixel-2 phone.
For aarch32:
- Before this change q8gemm-bench was 1.7% slower overall. After this it was .9%.
- convolution-bench was too noisy with wide variance so did not account for it.

For aarch64 also it was inconclusive.

Generally it was hard to get a good pulse at op level benchmarking though. I left the changes in largely because they seemed like a better schedule. In out-of-order processor it may not mean much though.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

### CODE REVIEW

**Review by dreiss — CHANGES_REQUESTED** [review]
Can you update your test plan to indicate what benchmarks you ran and what the results were.

**Comment by kimishpatel** [comment]
> Can you update your test plan to indicate what benchmarks you ran and what the results were.

@dreiss Updated comment.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S [hunk 1]
```diff
@@ -78,18 +78,30 @@ BEGIN_FUNCTION pytorch_q8gemm_ukernel_4x8__aarch32_neon
...
+    LDR r7, [sp, 44]
+    VPUSH {d8-d15}
+
+    # Load bias0123, bias4567
+    VLDM ip!, {d16-d19}
 
     # Load output channel index
-    LDR r5, [sp, 100]
+    LDR r5, [sp, 104]
+    CMP r0, 2
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S:93]: Where are you using the result of this comparison?  I don't seem to be able to spot it.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-aarch64-neon.S [hunk 2]
```diff
@@ -42,33 +42,37 @@ BEGIN_FUNCTION pytorch_q8gemm_ukernel_8x8__aarch64_neon
 
     # https://developer.arm.com/docs/ihi0055/d/procedure-call-standard-for-the-arm-64-bit-architecture
     # Callee need to save 8-15 vector registers and only the lower 64 bits of each.
+
+    # Load params
+    LDP x16, x8, [sp]
+
+    # a1
+    CMP x0, 2
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-aarch64-neon.S:50]: Same.  I would have expected a condition code on ADD if that was supposed to be conditional?

**Review by AshkanAliabadi — COMMENTED** [review]
Thanks. LGTM except for the two questions I have.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S:93]: This is from the original code. It is just movement of code that seem to have hidden its usage. It here: https://github.com/pytorch/pytorch/pull/37626/files/20a19b900e54e205185d5ce54a3ec327e54e12c2#diff-f682e6802902a89ec92f72efe5ec2a5bR120

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-aarch64-neon.S (hunk 2 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-aarch64-neon.S:50]: Same as earlier. Just hidden away. https://github.com/pytorch/pytorch/pull/37626/files/20a19b900e54e205185d5ce54a3ec327e54e12c2#diff-f682e6802902a89ec92f72efe5ec2a5bR120

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S (hunk 1 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S:93]: This is very dangerous.  Condition flags can be implicitly set as a side effect by quite a few instructions which makes putting too much distance between condition checks and their actual usage dangerous.  For instance if someone a month from now modifies this code to add a SUBS in between this  CMP and CSEL this code will break.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-aarch64-neon.S (hunk 2 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-aarch64-neon.S:50]: Likewise.  It's should be fine to move the ADD as far as I can tell, but moving the CMP is a hard to find bug in the making. :)

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S:93]: OK, so I think you could also say that if someone is changing this code then they probably understand what they are doing. I do understand if this code gets larger it become intractable but probably you wont write really long handwritten assembly. Having said all that, in this instance I dont think there is much justification to move the cmp around. I was trying to hide latency but I doubt it helps much. I did not really account the point you mentioned, so thats fair. I will move it back to where it was.

Comment (AshkanAliabadi) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S:93]: Again, as mentioned before, all my comments are merely suggestions and you are more than welcome to do as you see fit.

**Review by AshkanAliabadi — APPROVED** [review]

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S:93]: Made the changes.

**Pull request closed without merging**